Zend/Optimizer/zend_optimizer.c

```diff
@@ -831,6 +831,50 @@ zend_class_entry *zend_optimizer_get_class_entry_from_op1(
 	return NULL;
 }
 
+const zend_class_constant *zend_fetch_class_const_info(
+	const zend_script *script, const zend_op_array *op_array, const zend_op *opline, bool *is_prototype) {
+	const zend_class_entry *ce = NULL;
+	bool is_static_or_parent_reference = false;
+
+	if (opline->op2_type != IS_CONST || Z_TYPE(ZEND_OP2_LITERAL(opline)) != IS_STRING) {
+		return NULL;
+	}
+	if (opline->op1_type == IS_CONST) {
+		zval *op1 = CRT_CONSTANT(opline->op1);
+		if (Z_TYPE_P(op1) == IS_STRING) {
+			ce = zend_optimizer_get_class_entry(script, op_array, Z_STR_P(op1 + 1));
+		}
+	} else if (opline->op1_type == IS_UNUSED && op_array->scope && !(op_array->scope->ce_flags & ZEND_ACC_TRAIT)) {
+		int fetch_type = (int) opline->op1.num & ZEND_FETCH_CLASS_MASK;
+		if (fetch_type == ZEND_FETCH_CLASS_SELF) {
+			ce = op_array->scope;
+		} else if (fetch_type == ZEND_FETCH_CLASS_STATIC) {
+			ce = op_array->scope;
+			is_static_or_parent_reference = true;
+		} else if (fetch_type == ZEND_FETCH_CLASS_PARENT) {
+			if (op_array->scope->ce_flags & ZEND_ACC_LINKED) {
+				ce = op_array->scope->parent;
+			}
+			is_static_or_parent_reference = true;
+		}
+	}
+	if (!ce || (ce->ce_flags & ZEND_ACC_TRAIT)) {
+		return NULL;
+	}
+	zend_class_constant *const_info = zend_hash_find_ptr(&ce->constants_table, Z_STR(ZEND_OP2_LITERAL(opline)));
+	if (!const_info) {
+		return NULL;
+	}
+	if ((ZEND_CLASS_CONST_FLAGS(const_info) & ZEND_ACC_DEPRECATED)
+		|| ((ZEND_CLASS_CONST_FLAGS(const_info) & ZEND_ACC_PPP_MASK) != ZEND_ACC_PUBLIC && const_info->ce != op_array->scope)) {
+		return NULL;
+	}
+	*is_prototype = !is_static_or_parent_reference
+		|| (const_info->ce->ce_flags & ZEND_ACC_FINAL) || (ZEND_CLASS_CONST_FLAGS(const_info) & ZEND_ACC_FINAL);
```

Comment:
This should be an `&&` condition. The class const is only a prototype if the access is on `static` _and_ the class and class const are not final.

Reply:
Ohh, I totally misunderstood what prototype means. I thought it was the other way around: if something is prototype then it can be optimized (so it's not non-final or referenced via self:: or parent::). Then I'll invert my conditions